kolibri/core/assets/src/views/sync/SelectAddressModalGroup/useDevices.js

```diff
@@ -0,0 +1,179 @@
+  const fetchFailed = ref(false);
+  const hasFetched = ref(false);
+
+  async function doFetch() {
+    set(isFetching, true);
+    set(fetchFailed, false);
+    try {
+      let _devices = await fetchDevices(apiParams);
+      set(
+        devices,
+        _devices.map(d => reactive(d))
+      );
+
+      set(hasFetched, true);
+      set(isFetching, false);
+    } catch (e) {
+      set(fetchFailed, true);
+      set(isFetching, false);
+    }
+  }
+
+  // Start polling
+  const fetch = useTimeoutPoll(doFetch, 5000, { immediate: true });
+
+  // Stop polling
+  onBeforeUnmount(() => {
+    fetch.pause();
+  });
+
+  return {
+    devices,
+    isFetching,
+    fetchFailed,
+    hasFetched,
+    forceFetch: doFetch,
+  };
+}
+
+/**
+ *
+ * @param {{}} apiParams
+ * @param {function(NetworkLocation): Promise<bool>} filterFunction
+ * @return {{
+ *  devices: Ref<NetworkLocation[]>, hasFetched: Ref<bool>, isFetching: Ref<bool>,
+ *  filterFailed: Ref<bool>, forceFetch: (function(): Promise<void>)
+ * }}
+ */
+function useDevicesWithFilter(apiParams, filterFunction) {
+  const isFiltering = ref(false);
+  const filteringFailed = ref(false);
+  const hasFiltered = ref(false);
+  const availableIds = ref([]);
+  const unavailableIds = ref([]);
+  const { devices, isFetching, fetchFailed, hasFetched, forceFetch } = useDevices(apiParams);
+
+  const getIsAvailable = useMemoize(filterFunction, { getKey: device => device.id });
+
+  // await for changes in devices array
+  watch(devices, async devices => {
+    set(isFiltering, true);
+    let failed = false;
+
+    // Initiate 'is available' requests concurrently
+    await Promise.all(
+      get(devices).map(async device => {
+        try {
+          // result is memoized once successful
+          let isAvailable = await getIsAvailable(device);
+
+          // Put into refs to trigger reactive behavior in computed devices
+          if (isAvailable) {
+            if (get(availableIds).indexOf(device.id) < 0) {
+              get(availableIds).push(device.id);
+            }
+          } else {
+            if (get(unavailableIds).indexOf(device.id) < 0) {
+              get(unavailableIds).push(device.id);
+            }
+          }
+        } catch (e) {
+          // clear cache to try again on next poll
+          getIsAvailable.cache.delete(device.id);
+
+          // If 404, don't mark as failed
+          if (e.response.status !== 404) {
+            failed = true;
+          }
+        }
+      })
+    );
+
+    set(filteringFailed, failed);
+    set(isFiltering, false);
+    set(hasFiltered, true);
+  });
+
+  return {
+    // use computed array that depends on availableIds/unavailableIds
+    devices: computed(() => {
+      return get(devices)
+        .filter(d => get(unavailableIds).indexOf(d.id) < 0)
+        .map(d => {
+          // set unavailable if we haven't determined if it should be filtered out yet
+          d.available = get(availableIds).indexOf(d.id) >= 0;
```

Comment:
This was a fly-by decision to maintain visibility into the underlying processing, by showing locations as unavailable until we've determined the filtering result from `getIsAvailable`. If it takes a long time to call `getIsAvailable`, without this the user would see an empty list. The downside is that if the device gets filtered out, the device would vanish from the list.